design/service-indicator-lights.md

```diff
@@ -0,0 +1,201 @@
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks](#drawbacks)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 12 16:19:59 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+Many hardware platforms have "indicator" lights in the chassis front
+(or rear) panel and sometimes have other lights associated with
+individual components like drives inside the chassis. We need to
+provide a way to turn those lights on when service is needed for a
+host, and off when service is no longer needed. This document
+describes the new CRD and controller for managing indicator lights.
+
+## Motivation
+
+### Goals
+
+1. Enable operations staff to identify physical devices associated
+   with software-reported faults.
+
+### Non-Goals
+
+1. Actively manage all indicator lights for all known hardware.
+
+## Proposal
+
+### User Stories
+
+#### As an operator, I want to be able to match a software-reported fault to the hardware device that has failed.
+
+### Implementation Details/Notes/Constraints
+
+As with many BMC-related operators, we must take care not to overwhelm
+the management controller with too many requests for state or
+updates. Since a single chassis might have a dozen or more indicator
+lights, polling for their status regularly may place excessive load on
+the management controller.
+
+Unlike some other operations with BMCs, polling or changing the state
+of the indicator light for an internal component may require
+interacting with the I/O controller, which may interrupt regular I/O
+operations. For this reason, Ironic does not regularly poll the status
+of indicator lights and does not cache their status when an outside
+user queries it.
+
+We do not want to disturb production data flow any more than
+necessary, so we want to limit the amount of polling we do related to
+indicator lights from metal3.
+
+Because of these 2 constraints, we do not want to add indicator
+management to the existing BareMetalHost controller or CRD. Instead,
+we will define a new CRD to represent a service indicator light that
+is in use.
+
+Only lights associated with an instance of the new CRD will be managed
+at all, and that management is limited to checking and toggling the
+state of the light when the resource in k8s changes. No active
+management is performed to ensure that state continues to match the
+resource.
+
+### Risks and Mitigations
+
+1. Polling too frequently will cause more than the usual level of
+   issues with the BMC or disk controllers. We mitigate these problems
+   by reducing the number of devices we actually poll, and reducing
+   the polling frequency.
+
+2. Because less polling is performed, the state of the CR is more
+   likely to become out of sync with the hardware if some other
+   component is also manipulating the indicator light. We can mitigate
+   the effect of that by ensuring that the code that turns the light
```

Comment:
What is our goal here? To avoid turning ON the light if it's already ON? Would that actually increase the load on BMC (two operations might be required in some/most cases)?

Reply:
To avoid asking about the state of the light more than necessary.

Reply:
We could also just try to turn the light on/off as needed, and if we get an error then we could interpret that to decide whether it's really an error. Does the API give enough information for us to know that a light is already on if we try to turn it on? Or is the Ironic API itself already idempotent?

Reply:
ironic API allows for GET'ing current LED state as well as PUT'ing the new desired state. It is not an error if the desired state is the same as the current one.

As of the current implementation, ironic does not have any intermediate logic or maintained  state when it comes to GET/PUT the indicators. Conceptually, ironic client query traverses ironic and hits the BMC.

Reply:
If ironic is not going to consider it an error to be told to turn on a light that is already on, we don't need to query first from the operator.